sources/academy/webscraping/scraping_basics_python/08_saving_data.md

```diff
@@ -1,13 +1,246 @@
 ---
 title: Saving data with Python
 sidebar_label: Saving data
-description: TODO
+description: Lesson about building a Python application for watching prices. Using standard library to save data scraped from product listing pages in popular formats such as CSV or JSON.
 sidebar_position: 8
 slug: /scraping-basics-python/saving-data
 ---
 
-:::danger Work in progress
+**In this lesson, we'll save the data we scraped in the popular formats, such as CSV or JSON. We'll use Python's standard library to export the files.**
 
-This course is incomplete. As we work on adding new lessons, we would love to hear your feedback. You can comment right here under each page or [file a GitHub Issue](https://github.com/apify/apify-docs/issues) to discuss a problem.
+---
+
+We managed to scrape data about products and print it, with each product separated by a new line and each field separated by the `|` character. This already produces structured text that can be parsed, i.e., read programmatically.
+
+```text
+$ python main.py
+JBL Flip 4 Waterproof Portable Bluetooth Speaker | 74.95 | 74.95
+Sony XBR-950G BRAVIA 4K HDR Ultra HD TV | 1398.00 | None
+...
+```
+
+However, the format of this text is rather _ad hoc_ and does not adhere to any specific standard that others could follow. It's unclear what to do if a product title already contains the `|` character or how to represent multi-line product descriptions. No ready-made library can handle all the parsing.
+
+We should use widely popular formats that have well-defined solutions for all the corner cases and that other programs can read without much effort. Two such formats are CSV (_Comma-separated values_) and JSON (_JavaScript Object Notation_).
+
+## Collecting data
+
+Producing results line by line is an efficient approach to handling large datasets, but to simplify this lesson, we'll store all our data in one variable. This'll take three changes to our program:
+
+```py
+import httpx
+from bs4 import BeautifulSoup
+from decimal import Decimal
+
+url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+response = httpx.get(url)
+response.raise_for_status()
+
+html_code = response.text
+soup = BeautifulSoup(html_code, "html.parser")
+
+# highlight-next-line
+data = []
+for product in soup.select(".product-item"):
+    title = product.select_one(".product-item__title").text.strip()
```

Comment:
`select_one` can return also a `None`, but I believe it's fine for the purpose of the academy

Reply:
yup, I thought about this one, but opting to have the code shorter and fail fast if something's wrong